fog/sample-paykit/src/cached_tx_data/mod.rs

```diff
@@ -390,6 +460,127 @@ impl CachedTxData {
+        // indices.
+        let mut missed_block_indices =
+            CachedTxData::create_missed_block_indices(missed_block_ranges);
+
+        for block_datum in block_data {
+            missed_block_indices.remove(&block_datum.index);
+        }
+
+        // Sort the indices in order to create the fewest amount of block
+        // ranges possible in the iteration step below.
+        let mut missed_block_indices_sorted: Vec<u64> =
+            missed_block_indices.into_iter().collect::<Vec<_>>();
+        missed_block_indices_sorted.sort_unstable();
+
+        let mut updated_missed_block_ranges: Vec<common::BlockRange> = Vec::new();
+
+        if missed_block_indices_sorted.is_empty() {
+            return updated_missed_block_ranges;
+        }
+
+        if missed_block_indices_sorted.len() == 1 {
+            let first_block_index = missed_block_indices_sorted[0];
+            updated_missed_block_ranges.push(common::BlockRange {
+                start_block: first_block_index,
+                end_block: first_block_index + 1,
+            });
+            return updated_missed_block_ranges;
+        }
+
+        let mut start_block_index = missed_block_indices_sorted[0];
+        let mut i: usize = 0;
+        let mut j: usize = 1;
+        while j < missed_block_indices_sorted.len() {
+            let first_block_index = missed_block_indices_sorted[i];
+            let second_block_index = missed_block_indices_sorted[j];
+            // We've found a gap in indices, so the current range is complete
+            // and should be added.
+            if second_block_index - first_block_index > 1 {
+                let block_range = common::BlockRange {
+                    start_block: start_block_index,
+                    end_block: first_block_index + 1,
+                };
+                updated_missed_block_ranges.push(block_range);
+                start_block_index = second_block_index;
+            }
+
+            i += 1;
+            j += 1;
+        }
+
+        let last_block_index = missed_block_indices_sorted.last().unwrap();
+        let last_block_range = common::BlockRange {
+            start_block: start_block_index,
+            end_block: last_block_index + 1,
+        };
+
+        updated_missed_block_ranges.push(last_block_range);
+
+        updated_missed_block_ranges
+    }
+
+    fn create_missed_block_indices(missed_block_ranges: &[common::BlockRange]) -> HashSet<u64> {
+        let mut missed_block_indices: HashSet<u64> = HashSet::new();
+        for missed_block_range in missed_block_ranges {
+            for missed_block_index in missed_block_range.start_block..missed_block_range.end_block {
+                missed_block_indices.insert(missed_block_index);
+            }
+        }
+
+        missed_block_indices
+    }
+
+    // Converts a ledger::BlockResponses to a Vec<TxOutRecord>.
+    fn create_tx_out_records(block_response: &ledger::BlockResponse) -> Vec<TxOutRecord> {
```

Comment:
I might be missing something, but shouldn't this do a view key scan to see if the TxOut in the block actually belongs to the account?

Reply:
I think tests will still pass but we'll end up with a bunch of "View key scanning failed, fog gave us a TXO that wasn't ours" logs?